File: python-version/lrs_tools/topology.py
"""Gap, overlap, neighbor, and validation helpers for LRS event tables."""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np
import pandas as pd

from lrs_tools.schema import LrsSchema, coerce_numeric, pad_roadway_id


def _prepared(df: pd.DataFrame, schema: LrsSchema) -> pd.DataFrame:
    work = df.copy()
    work[schema.bmp] = coerce_numeric(work[schema.bmp])
    work[schema.emp] = coerce_numeric(work[schema.emp])
    work["_merge_roadway"] = work[schema.roadway].map(pad_roadway_id)
    return work.sort_values(by=["_merge_roadway", schema.bmp, schema.emp]).reset_index(drop=True)


def find_gaps(df: pd.DataFrame, schema: LrsSchema | None = None) -> pd.DataFrame:
    """Return coverage gaps between consecutive events on the same roadway."""
    if df.empty:
        return pd.DataFrame(columns=["ROADWAY", "GAP_BMP", "GAP_EMP", "GAP_LENGTH"])

    schema = schema or LrsSchema.from_dataframe(df, require_line=True)
    schema.validate_line(df)
    work = _prepared(df, schema)

    prev_road = work["_merge_roadway"].shift(1)
    prev_emp = work[schema.emp].shift(1)
    same_road = work["_merge_roadway"].notna() & (work["_merge_roadway"] == prev_road)
    gap_len = work[schema.bmp] - prev_emp
    is_gap = same_road & gap_len.notna() & (gap_len > schema.tolerance)

    gaps = work.loc[is_gap, ["_merge_roadway", schema.bmp]].copy()
    gaps = gaps.rename(columns={"_merge_roadway": "ROADWAY", schema.bmp: "GAP_EMP"})
    gaps["GAP_BMP"] = prev_emp[is_gap].values
    gaps["GAP_LENGTH"] = gaps["GAP_EMP"] - gaps["GAP_BMP"]
    return gaps[["ROADWAY", "GAP_BMP", "GAP_EMP", "GAP_LENGTH"]].reset_index(drop=True)


def find_overlaps(df: pd.DataFrame, schema: LrsSchema | None = None) -> pd.DataFrame:
    """Return pairwise overlaps of events on the same roadway."""
    schema = schema or LrsSchema.from_dataframe(df, require_line=True)
    schema.validate_line(df)
    work = _prepared(df, schema)
    work = work.dropna(subset=["_merge_roadway", schema.bmp, schema.emp])
    if work.empty:
        return pd.DataFrame(columns=["ROADWAY", "LEFT_BMP", "LEFT_EMP", "RIGHT_BMP", "RIGHT_EMP"])

    rows: list[dict] = []
    for road, group in work.groupby("_merge_roadway", sort=False):
        records = group[[schema.bmp, schema.emp]].to_dict("records")
        for i, left in enumerate(records):
            for right in records[i + 1 :]:
                if left[schema.bmp] < right[schema.emp] and left[schema.emp] > right[schema.bmp]:
                    rows.append(
                        {
                            "ROADWAY": road,
                            "LEFT_BMP": left[schema.bmp],
                            "LEFT_EMP": left[schema.emp],
                            "RIGHT_BMP": right[schema.bmp],
                            "RIGHT_EMP": right[schema.emp],
                        }
                    )
    return pd.DataFrame(rows)


def neighbors_along_route(
    df: pd.DataFrame,
    value_col: str,
    schema: LrsSchema | None = None,
    *,
    sort_col: str | None = None,
) -> pd.DataFrame:
    """Attach upstream/downstream neighbor values along each roadway."""
    if value_col not in df.columns:
        raise ValueError(f"value_col '{value_col}' is not in the DataFrame")

    work = df.copy()
    if schema is None:
        schema = LrsSchema.from_dataframe(work)
    if schema.roadway not in work.columns:
        raise ValueError(f"roadway column '{schema.roadway}' is not in the DataFrame")

    order_col = sort_col or schema.measure or schema.bmp
    if order_col not in work.columns:
        raise ValueError(f"sort column '{order_col}' is not in the DataFrame")

    work[order_col] = coerce_numeric(work[order_col])
    work["_merge_roadway"] = work[schema.roadway].map(pad_roadway_id)
    work = work.sort_values(by=["_merge_roadway", order_col]).reset_index(drop=True)
    grouped = work.groupby("_merge_roadway", dropna=False)

    work["Upstream_Value"] = grouped[value_col].shift(1)
    work["Downstream_Value"] = grouped[value_col].shift(-1)
    work["Upstream_Measure"] = grouped[order_col].shift(1)
    work["Downstream_Measure"] = grouped[order_col].shift(-1)

    up_ok = work[value_col].notna() & work["Upstream_Value"].notna() & (work["Upstream_Value"] != 0)
    down_ok = work[value_col].notna() & work["Downstream_Value"].notna() & (work["Downstream_Value"] != 0)
    work["Upstream_Diff_pct"] = np.where(
        up_ok, (work[value_col] - work["Upstream_Value"]) / work["Upstream_Value"], np.nan
    )
    work["Downstream_Diff_pct"] = np.where(
        down_ok, (work[value_col] - work["Downstream_Value"]) / work["Downstream_Value"], np.nan
    )
    return work.drop(columns=["_merge_roadway"])


@dataclass
class LrsValidation:
    invalid_bounds: pd.DataFrame
    overlaps: pd.DataFrame
    null_keys: pd.DataFrame
    gaps: pd.DataFrame

    @property
    def ok(self) -> bool:
        return self.invalid_bounds.empty and self.overlaps.empty and self.null_keys.empty


def validate_lrs(df: pd.DataFrame, schema: LrsSchema | None = None) -> LrsValidation:
    """Flag BMP >= EMP, null keys, and overlapping events on the same roadway."""
    schema = schema or LrsSchema.from_dataframe(df, require_line=True)
    schema.validate_line(df)
    work = _prepared(df, schema)
    invalid_bounds = work[work[schema.bmp].notna() & work[schema.emp].notna() & (work[schema.bmp] >= work[schema.emp])].copy()
    null_keys = work[work["_merge_roadway"].isna() | work[schema.bmp].isna() | work[schema.emp].isna()].copy()
    overlaps = find_overlaps(work.drop(columns=["_merge_roadway"], errors="ignore"), schema)
    gaps = find_gaps(work.drop(columns=["_merge_roadway"], errors="ignore"), schema)
    return LrsValidation(
        invalid_bounds=invalid_bounds.drop(columns=["_merge_roadway"], errors="ignore"),
        overlaps=overlaps,
        null_keys=null_keys.drop(columns=["_merge_roadway"], errors="ignore"),
        gaps=gaps,
    )

File: js-version/js/topology.js
(function (root) {
  const isNode = typeof module === "object" && module.exports;
  const LRS = isNode ? require("./schema.js") : root.LRS || (root.LRS = {});

  function prepared(rows, schema) {
    const work = LRS.cloneRows(rows);
    work.forEach((row) => {
      row[schema.bmp] = LRS.coerceNumeric(row[schema.bmp]);
      row[schema.emp] = LRS.coerceNumeric(row[schema.emp]);
      row._merge_roadway = LRS.padRoadwayId(row[schema.roadway]);
    });
    return LRS.sortRows(work, ["_merge_roadway", schema.bmp, schema.emp]);
  }

  function extraGroupCols(rows, schema, extraCols) {
    const columns = LRS.columnsOf(rows);
    return (extraCols || []).filter((col) => col && columns.includes(col) && col !== schema.roadway);
  }

  function extraGroupKey(row, extraCols) {
    if (!extraCols.length) return "";
    return extraCols.map((col) => String(LRS.compareKey(row[col]))).join("\u001f");
  }

  function findGaps(rows, schema, options = {}) {
    if (!rows.length) {
      return [];
    }
    schema = schema || LRS.LrsSchema.fromRows(rows, { requireLine: true });
    schema.validateLine(rows);
    const extraCols = extraGroupCols(rows, schema, options.groupCols);
    let work = prepared(rows, schema);
    work = LRS.sortRows(work, ["_merge_roadway", ...extraCols, schema.bmp, schema.emp]);
    const gaps = [];
    for (let i = 1; i < work.length; i += 1) {
      const prev = work[i - 1];
      const curr = work[i];
      const sameGroup =
        curr._merge_roadway != null &&
        curr._merge_roadway === prev._merge_roadway &&
        extraGroupKey(curr, extraCols) === extraGroupKey(prev, extraCols);
      const gapLen = curr[schema.bmp] != null && prev[schema.emp] != null
        ? curr[schema.bmp] - prev[schema.emp]
        : null;
      if (sameGroup && gapLen != null && gapLen > schema.tolerance) {
        gaps.push({
          ROADWAY: curr._merge_roadway,
          GAP_BMP: prev[schema.emp],
          GAP_EMP: curr[schema.bmp],
          GAP_LENGTH: curr[schema.bmp] - prev[schema.emp],
        });
      }
    }
    return gaps;
  }

  function findOverlaps(rows, schema, options = {}) {
    schema = schema || LRS.LrsSchema.fromRows(rows, { requireLine: true });
    schema.validateLine(rows);
    const extraCols = extraGroupCols(rows, schema, options.groupCols);
    let work = prepared(rows, schema).filter(
      (row) => row._merge_roadway != null && row[schema.bmp] != null && row[schema.emp] != null
    );
    const out = [];
    const groups = LRS.groupBy(work, (row) => `${row._merge_roadway}\u001f${extraGroupKey(row, extraCols)}`);
    for (const [key, group] of groups.entries()) {
      const road = key.split("\u001f")[0];
      for (let i = 0; i < group.length; i += 1) {
        for (let j = i + 1; j < group.length; j += 1) {
          const left = group[i];
          const right = group[j];
          if (left[schema.bmp] < right[schema.emp] && left[schema.emp] > right[schema.bmp]) {
            out.push({
              ROADWAY: road,
              LEFT_BMP: left[schema.bmp],
              LEFT_EMP: left[schema.emp],
              RIGHT_BMP: right[schema.bmp],
              RIGHT_EMP: right[schema.emp],
            });
          }
        }
      }
    }
    return out;
  }

  function neighborsAlongRoute(rows, valueCol, options = {}) {
    const columns = LRS.columnsOf(rows);
    if (!columns.includes(valueCol)) {
      throw new Error(`value_col '${valueCol}' is not in the DataFrame`);
    }
    const work = LRS.cloneRows(rows);
    const schema = options.schema || LRS.LrsSchema.fromRows(work);
    if (!columns.includes(schema.roadway)) {
      throw new Error(`route ID column '${schema.roadway}' is not in the DataFrame`);
    }
    const orderCol = options.sortCol || schema.measure || schema.bmp;
    if (!columns.includes(orderCol)) {
      throw new Error(`sort column '${orderCol}' is not in the DataFrame`);
    }
    work.forEach((row) => {
      row[orderCol] = LRS.coerceNumeric(row[orderCol]);
      row._merge_roadway = LRS.padRoadwayId(row[schema.roadway]);
    });
    const sorted = LRS.sortRows(work, ["_merge_roadway", orderCol]);
    const groups = LRS.groupBy(sorted, (row) => row._merge_roadway);
    const result = [];
    for (const group of groups.values()) {
      for (let i = 0; i < group.length; i += 1) {
        const row = LRS.dropColumns(group[i], ["_merge_roadway"]);
        const prev = i > 0 ? group[i - 1] : null;
        const next = i < group.length - 1 ? group[i + 1] : null;
        row.Upstream_Value = prev ? prev[valueCol] : null;
        row.Downstream_Value = next ? next[valueCol] : null;
        row.Upstream_Measure = prev ? prev[orderCol] : null;
        row.Downstream_Measure = next ? next[orderCol] : null;
        const value = LRS.coerceNumeric(row[valueCol]);
        const up = LRS.coerceNumeric(row.Upstream_Value);
        const down = LRS.coerceNumeric(row.Downstream_Value);
        row.Upstream_Diff_pct = value != null && up != null && up !== 0 ? (value - up) / up : null;
        row.Downstream_Diff_pct =
          value != null && down != null && down !== 0 ? (value - down) / down : null;
        result.push(row);
      }
    }
    return result;
  }

  function validateLrs(rows, schema, options = {}) {
    schema = schema || LRS.LrsSchema.fromRows(rows, { requireLine: true });
    schema.validateLine(rows);
    const work = prepared(rows, schema);
    const invalidBounds = work.filter(
      (row) => row[schema.bmp] != null && row[schema.emp] != null && row[schema.bmp] > row[schema.emp]
    );
    const zeroLength = work.filter(
      (row) => row[schema.bmp] != null && row[schema.emp] != null && row[schema.bmp] === row[schema.emp]
    );
    const nullKeys = work.filter(
      (row) => row._merge_roadway == null || row[schema.bmp] == null || row[schema.emp] == null
    );
    const cleaned = work.map((row) => LRS.dropColumns(row, ["_merge_roadway"]));
    return {
      invalidBounds: invalidBounds.map((row) => LRS.dropColumns(row, ["_merge_roadway"])),
      zeroLength: zeroLength.map((row) => LRS.dropColumns(row, ["_merge_roadway"])),
      overlaps: findOverlaps(cleaned, schema, options),
      nullKeys: nullKeys.map((row) => LRS.dropColumns(row, ["_merge_roadway"])),
      gaps: findGaps(cleaned, schema, options),
      get ok() {
        return (
          !this.invalidBounds.length &&
          !this.zeroLength.length &&
          !this.overlaps.length &&
          !this.nullKeys.length &&
          !this.gaps.length
        );
      },
    };
  }

  Object.assign(LRS, { findGaps, findOverlaps, neighborsAlongRoute, validateLrs });
  if (isNode) module.exports = LRS;
})(typeof globalThis !== "undefined" ? globalThis : this);
